src/Microsoft.ML.Data/Training/TrainerEstimatorBase.cs

```diff
@@ -41,6 +41,11 @@ public abstract class TrainerEstimatorBase<TTransformer, TModel> : ITrainerEstim
         /// </summary>
         public readonly SchemaShape.Column WeightColumn;
 
+        /// <summary>
+        /// The optional groupID column that the ranking trainers expects.
+        /// </summary>
+        public readonly SchemaShape.Column GroupIdColumn;
```

Comment:
>GroupIdColumn [](start = 43, length = 13)

I am not sure I like the idea that every estimator will have the group ID column, but it only makes sense for ranking trainers.

Reply:
The alternative would be to modify the role mapped data after the  estimator constructs it, and inject the groupId, and that feels error-prone. It's just one reference :)

---
In reply to: [227866406](https://github.com/dotnet/machinelearning/pull/1352#discussion_r227866406) [](ancestors = 227866406)

Reply:
So we have a `TrainerEstimatorBase`, is there anything that prevents us from having another subclass, if we wanted to have *additional* roles? We may have to change something below to be a virtual method or have some other extension point, but I feel like this would be helpful anyway.

I am not really a fan of having a single utility base class handle everything, since if it is capable of doing *everything* I suspect it will become too unwieldly.

---
In reply to: [227895594](https://github.com/dotnet/machinelearning/pull/1352#discussion_r227895594) [](ancestors = 227895594,227866406)